README.md
# DMemu 

 :blue_book: <a href = "https://dmemu.readthedocs.io/en/latest/">Documentation</a>

`DMemu` is a python package implementing nonlinear response of different dark matter extension of $\Lambda \rm CDM$ model. Using a fitting function or an emulator, a nonlinear $\Lambda \rm CDM$ matter power spectrum is modified to accommodate a nonlinear matter power spectrum of a specific $\Lambda \rm CDM$ extension as resulting from gravity-only $N$-body simulations. Currently included emulators are two-body decaying dark matter (`TBDemu`).

## TBDemu

## Describtion

`TBDemu` is an emulator implementing a nonlinear response of two-body decays within the dark matter [[1]](#1). The phenomenology of two-body decaying dark matter (2bDDM) is based on two parameters: the decay rate $\Gamma$ (in 1/Gyr) and the magnitude of velocity kicks obtained by decay products $v_k$ (in km/s). Additionally, one can assume only a fraction of decaying dark matter $f$ in the total dark matter abundance: $f=\Omega_{\rm m, decaying}/\Omega_{\rm m, total}$. The `TBDemu` emulator is built on gravity-only $N$-body simulations run by `Pkdgrav3` code [[2]](#2). The emulator predicts

<p>$\mathcal{S}^{\Gamma,v_k,f}_{\rm DDM}(k,z) = P^{\Gamma,v_k,f}_{\rm DDM}(k,z)/P_{\Lambda \rm CDM}(k,z)$,</p>

thus a ratio of nonlinear matter power spectrum in the scenario including dark matter decays and (nonlinear) $\Lambda \rm CDM$ matter power spectrum. For more details, see Ref. [[1]](#1). The emulator was trained using *Sinusoidal representation networks* (SIRENs) [[3]](#3).

## Installation

Clone the git repository:

```bash
git clone git@github.com:jbucko/TBDemu.git
```

Then navigate to the root folder of the repository and install the package using a `setup.py` file:

```bash
python setup.py install
```

The requirements for the package are listed in `requirements.txt` and will be installed automatically (if not already installed).

## Usage

```python
import numpy as np
import DMemu
import matplotlib.pyplot as plt

# load emulator
emul = DMemu.TBDemu()

# predict suppressions between kmin and kmax for a single redshift
kmin = 1e-3 # in h/Mpc
kmax = 5 # in h/Mpc
ks = np.logspace(np.log10(kmin),np.log10(kmax),200) # scales
zs = 0.0 # redshift
velocity_kick = 500 # in km/s
gamma_decay = 1/50 # in 1/Gyr
fraction = 1.0

pks = emul.predict(ks,zs,fraction,velocity_kick,gamma_decay)

# plot
plt.semilogx(ks,pks)
plt.xlabel(r'$k$ [h/Mpc]')
plt.ylabel(r'$P_{\rm DDM}/P_{\Lambda \rm CDM}$')
plt.tight_layout()
plt.show()
```

<img src='examples/tbdemu_example.jpeg' width='600'/>

### Parameter space:
* decay rate: $\Gamma \in [0,1/13.5]$ Gyr $^{-1}$  
* velocity kick magnitude: $v_k \in [0,5000]$ km/s  
* fraction of 2bDDM: $f \in [0,1]$  
* scales: $k < 6$ h/Mpc  
* redshifts: $z < 2.35$

### Input format of $k$ and $z$
1.
    Single value of $k$ and $z$:
    
    ```python
    k = 0.10 # in h/Mpc
    z = 0.0
    pks = emul.predict(k,z,fraction,velocity_kick,gamma_decay)
    ```
    Provides a single suppression value.

2.
    Single value of $z$ for multiple scales $k$:
    
    ```python
    k = np.logspace(-2,0,10) # in h/Mpc
    z = 0.0
    pks = emul.predict(k,z,fraction,velocity_kick,gamma_decay)
    ```
    Provides a list of suppressions at desired scales for a single redshift $z$.

3.
    Single value of $k$ for multiple redshifts $z$:
    
    ```python
    k = 0.10 # in h/Mpc
    z = np.array([0.0,1.0,2.0])
    pks = emul.predict(k,z,fraction,velocity_kick,gamma_decay)
    ```
    Provides a list of suppressions at a given scale for all redshift values $z$.

4.
    Multiple scales $k$ for multiple redshifts $z$:
    
    ```python
    k = np.array([0.1,0.5,1.0]) # in h/Mpc
    z = np.array([0.0,1.0,2.0])
    pks = emul.predict(k,z,fraction,velocity_kick,gamma_decay)
    ```
    The above code provides three suppression values, first for $k=0.1$ and $z=0.0$, second for $k=0.5$ and $z=1.0$ and last for $k=1.0$ and $z=2.0$. The code checks that the lengths of both array are equal.
    
## Extrapolation

Extrapolation for $\Gamma$, $v_k$, $f$ and $z$ is not allowed as the trained architecture cannot reliably predict outside the training domain. Extrapolation for $k>6$ h/Mpc is done by adding a constant suppression continuously attached to the one provided by an emulator, see the figure below. 

<img src='examples/tbdemu_extrapolation.jpeg' width='600'/>

## References
<a id="1">[1]</a> 
Bucko et al. 2023, in prep.

<a id="2">[2]</a> 
Potter, D., Stadel, J. & Teyssier, R. PKDGRAV3: beyond trillion particle cosmological simulations for the next era of galaxy surveys. Comput. Astrophys. 4, 2 (2017). https://doi.org/10.1186/s40668-017-0021-1

<a id="3">[3]</a> 
Sitzmann, V., Martel, J. N. P., Bergman, A. W., Lindell, D. B., & Wetzstein, G. (2020). Implicit Neural Representations with Periodic Activation Functions. Proc. NeurIPS.


docs/changelog.rst
=========
Changelog
=========

v1.0
----
* Initial form of the package
* Compatible with python 3 only 


src/my_lib_colab5.py

import torch
from torch import nn
import torch.nn.functional as F
import numpy as np

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')


class SineLayer(nn.Module):
    # See paper sec. 3.2, final paragraph, and supplement Sec. 1.5 for discussion of omega_0.
    
    # If is_first=True, omega_0 is a frequency factor which simply multiplies the activations before the 
    # nonlinearity. Different signals may require different omega_0 in the first layer - this is a 
    # hyperparameter.
    
    # If is_first=False, then the weights will be divided by omega_0 so as to keep the magnitude of 
    # activations constant, but boost gradients to the weight matrix (see supplement Sec. 1.5)
    
    def __init__(self, in_features, out_features, bias=True,
                 is_first=False, omega_0=30):
        super().__init__()
        self.omega_0 = omega_0
        self.is_first = is_first
        
        self.in_features = in_features
        self.linear = nn.Linear(in_features, out_features, bias=bias)
        
        self.init_weights()
    
    def init_weights(self):
        with torch.no_grad():
            if self.is_first:
                self.linear.weight.uniform_(-1 / self.in_features, 
                                             1 / self.in_features)      
            else:
                self.linear.weight.uniform_(-np.sqrt(6 / self.in_features) / self.omega_0, 
                                             np.sqrt(6 / self.in_features) / self.omega_0)
        
    def forward(self, input):
        return torch.sin(self.omega_0 * self.linear(input))
    
    def forward_with_intermediate(self, input): 
        # For visualization of activation distributions
        intermediate = self.omega_0 * self.linear(input)
        return torch.sin(intermediate), intermediate
    
    
class Siren(nn.Module):
    def __init__(self, in_features, hidden_features, hidden_layers, out_features, outermost_linear=False, 
                 first_omega_0=30, hidden_omega_0=30.):
        super().__init__()
        
        self.net = []
        self.net.append(SineLayer(in_features, hidden_features, 
                                  is_first=True, omega_0=first_omega_0))

        for i in range(hidden_layers):
            self.net.append(SineLayer(hidden_features, hidden_features, 
                                      is_first=False, omega_0=hidden_omega_0))

        if outermost_linear:
            final_linear = nn.Linear(hidden_features, out_features)
            
            with torch.no_grad():
                final_linear.weight.uniform_(-np.sqrt(6 / hidden_features) / hidden_omega_0, 
                                              np.sqrt(6 / hidden_features) / hidden_omega_0)
                
            self.net.append(final_linear)
        else:
            self.net.append(SineLayer(hidden_features, out_features, 
                                      is_first=False, omega_0=hidden_omega_0))
        
        self.net = nn.Sequential(*self.net)
    
    def forward(self, coords):
        coords = coords.clone().detach().requires_grad_(True) # allows to take derivative w.r.t. input
        output = self.net(coords)
        return output, coords        

    def forward_with_activations(self, coords, retain_grad=False):
        '''Returns not only model output, but also intermediate activations.
        Only used for visualizing activations later!'''
        activations = OrderedDict()

        activation_count = 0
        x = coords.clone().detach().requires_grad_(True)
        activations['input'] = x
        for i, layer in enumerate(self.net):
            if isinstance(layer, SineLayer):
                x, intermed = layer.forward_with_intermediate(x)
                
                if retain_grad:
                    x.retain_grad()
                    intermed.retain_grad()
                    
                activations['_'.join((str(layer.__class__), "%d" % activation_count))] = intermed
                activation_count += 1
            else: 
                x = layer(x)
                
                if retain_grad:
                    x.retain_grad()
                    
            activations['_'.join((str(layer.__class__), "%d" % activation_count))] = x
            activation_count += 1

        return activations

class LinearLayer(nn.Module):
    # See paper sec. 3.2, final paragraph, and supplement Sec. 1.5 for discussion of omega_0.
    
    # If is_first=True, omega_0 is a frequency factor which simply multiplies the activations before the 
    # nonlinearity. Different signals may require different omega_0 in the first layer - this is a 
    # hyperparameter.
    
    # If is_first=False, then the weights will be divided by omega_0 so as to keep the magnitude of 
    # activations constant, but boost gradients to the weight matrix (see supplement Sec. 1.5)
    
    def __init__(self, in_features, out_features, bias=True,
                 is_first=False, omega_0=1.0):
        super().__init__()
        self.omega_0 = omega_0
        self.is_first = is_first
        
        self.in_features = in_features
        self.linear = nn.Linear(in_features, out_features, bias=bias)
        # self.activation = nn.LeakyReLU(0.01)
        # self.activation = nn.Softmax()
        self.activation = nn.ReLU()
        
        self.init_weights()
    
    def init_weights(self):
        with torch.no_grad():
            if self.is_first:
                self.linear.weight.uniform_(-1 / self.in_features, 
                                             1 / self.in_features)      
            else:
                self.linear.weight.uniform_(-np.sqrt(1 / self.in_features) / self.omega_0, 
                                             np.sqrt(1 / self.in_features) / self.omega_0) # original sqrt(6/...)
        
    def forward(self, input):
        return self.activation(self.omega_0 * self.linear(input))
    



class relu_net(nn.Module):
    def __init__(self,num_input,num_output,hidden):
      super(relu_net, self).__init__()
      
      self.hidden = hidden
      self.input = nn.Linear(num_input, self.hidden[0])
      self.output = nn.Linear(self.hidden[-1], num_output)
      self.hidden_arr = []
      for i in range(len(self.hidden) - 1):
        self.hidden_arr.append(nn.Linear(self.hidden[i], self.hidden[i+1]))

    def l_hidden(self,nin,nout):
        return nn.Linear(nin,nout)

    # x represents our data
    def forward(self, x):
      x = self.input(x)
      x = F.relu(x).double()
      for i in range(len(self.hidden) - 1):
        # x = self.hidden_arr[i](x.double()).double()
        # l = self.l_hidden(self.hidden[i],self.hidden[i+1])
        # x = l(x)
        x = nn.Linear(self.hidden[i],self.hidden[i+1]).to(device)(x.float().to(device))
        x=F.relu(x)
        print(x.device)
      out = self.output(x.double())
      print(out.device)

      return out, x




src/TwoBodyDecaysEmulator.py
from typing import List
import numpy as np
import torch
import bisect
import os,sys
sys.path.append(os.path.dirname(__file__))
from my_lib_colab5 import *

class emulator():
    def __init__(self):
        PATH_TO_EMULATOR = os.path.dirname(__file__) + '/files/'
        self.emulator_DDM_2body = torch.load(PATH_TO_EMULATOR + 'emulator3D_bohb_optimal.pth',map_location=torch.device('cpu'))
        self.t_means = torch.tensor(np.genfromtxt(PATH_TO_EMULATOR+'PCA_means.csv'))
        self.t_eigenvectors = torch.tensor(np.genfromtxt(PATH_TO_EMULATOR+'PCA_eigenvectors.csv'))

    def check_interpret_k_z_input(self,k: float|list, z: float|list):
        # print('k,z types:',type(z) ,type(k))
        single_redshift = False
        if type(k) == float and type(z) == float:
            k = np.array([k])
            z = np.array([z])
            single_redshift = True
        elif type(z) == float and type(k) != float:
            if type(k) == list:
                k = np.array(k)
            z = np.array([z])
            single_redshift = True
        elif type(k) == float and type(z) != float:
            if type(z) == list:
                z = np.array(z)
            k_val = k
            k = np.array([k_val for _ in range(len(z))])
        elif type(k) != float and type(z) != float:
            if len(k) != len(z):
                raise Exception("Dimensions of k and z do not match:",len(k),len(z))
        return k,z,single_redshift
            
            
        

    def predict(self, k: float|list, z: float|list, f: float = 1.0, vk: float = 0., Gamma: float = 1e-10, p: float = 0.0) -> list:
        
        k,z,single_redshift = self.check_interpret_k_z_input(k,z)
        
        #defaults, do not change
        kmin = 1e-3
        kmax = 5.9
        nsteps_emul = 300
        # k_int = np.logspace(np.log10(kmin),np.log10(kmax),nsteps_emul)
        if max(z)>2.35: raise Exception("Too high redshift encountered (>2.35). Aborting...")
        if vk<0 or vk>5000: raise Exception("Velocity kick outside the training domain [0-5000] km/s. Aborting...")
        if Gamma<0 or Gamma>1/13.5: raise Exception("Decay rate outside the training domain [0-1/13.5] 1/Gyr. Aborting...")
        if f<0.0 or f>1.0: raise Exception("Fraction of 2bDDM outside the training domain [0-1]. Aborting...")

        
        kk = 105
        delta = np.log10(kmax/kmin)/(300-1)
        N = int(1+ np.log10(kk/kmin)/delta)
        k_ext = np.logspace(np.log10(kmin),np.log10(kmin) + (N-1)*delta,N)

        
        PP = []

        par_extend = np.tile(np.array([f,transform_v(vk),transform_g(Gamma)]),(len(z),1))
        params_concat = np.concatenate((par_extend,transform_z(z).reshape(-1,1)),axis = 1)

        
        with torch.no_grad():
            model_out_int_nPca, _ = self.emulator_DDM_2body(torch.tensor(params_concat))
        

        Pks_int = self.t_means + torch.matmul(model_out_int_nPca,self.t_eigenvectors)
        Pks_int = torch.clamp(Pks_int,0.,1.).numpy()

        Pks_ext = np.zeros((len(z),N))
        Pks_ext[:,:nsteps_emul] = Pks_int
        # print('Pk_int.shape,Pk_ext.shape:',Pks_int.shape,Pks_ext.shape)
        """
        p: value
        """
        column = Pks_int[:,-1]+(1-Pks_int[:,-1])*p
        extenstion = np.tile(column.reshape(-1,1),(1,N - nsteps_emul))
        Pks_ext[:,nsteps_emul:] = extenstion

        # for tracking of the extrapolation warning
        warning = False
        for i in range(len(k)):
            if k[i]<= kmin: PP.append(1.0) # for very large scales, boost is always 1.0
            else:
                z_idx = 0 if single_redshift == True else i # only one row in Pks_int and Pks_ext
                if k[i]>kmax: 
                    PP.append(column[z_idx])
                    if not warning: warning = True # extrapolation warning will be thrown before return
                else:
                    idx = bisect.bisect_left(k_ext,k[i])
                    lam = np.log10(k[i]/k_ext[idx-1])/np.log10(k_ext[idx]/k_ext[idx-1])
                    Pk_fill = (1-lam)*Pks_ext[z_idx,idx-1] + lam*Pks_ext[z_idx,idx]
                    PP.append(Pk_fill)
        if warning: print('[emulator.predict] Warning: Extrapolation required! Make sure you know what is hapenning.')
        return np.array(PP)


def transform_v(v: float) -> float:
    vks = [0,200,500,1000,5000]
    tvks = np.log10(np.array(vks)+50)
    tvmin = tvks[0]
    tvks -= tvmin
    tvmax = tvks[-1]
    return (np.log10(v + 50) - tvmin)/tvmax


def transform_g(g: float,a: float =115.396) -> float:
    # return g
    return np.log10(a*g+1)


def transform_z(z: float,a: float =3.97) -> float:
    # return z/zs[0]
    return np.log10(a*z+1)


def check_suppression(line: list) -> list:
    for i in range(len(line)):
        if line[i]>1.00:
            line[i] = 1.00
    return line 

src/DarkMatterEmulators.py
from typing import List
import numpy as np
import torch
import bisect
import os,sys
sys.path.append(os.path.dirname(__file__))
from my_lib_colab5 import *
from typing import Union

class TBDemu():
    def __init__(self):
        PATH_TO_EMULATOR = os.path.dirname(__file__) + '/files/'
        self.emulator_DDM_2body = torch.load(PATH_TO_EMULATOR + 'emulator3D_bohb_optimal.pth',map_location=torch.device('cpu'))
        self.t_means = torch.tensor(np.genfromtxt(PATH_TO_EMULATOR+'PCA_means.csv'))
        self.t_eigenvectors = torch.tensor(np.genfromtxt(PATH_TO_EMULATOR+'PCA_eigenvectors.csv'))

    def check_interpret_k_z_input(self,k: Union[float,list], z: Union[float,list]):
        # print('k,z types:',type(z) ,type(k))
        single_redshift = False
        if type(k) == float and type(z) == float:
            k = np.array([k])
            z = np.array([z])
            single_redshift = True
        elif type(z) == float and type(k) != float:
            if type(k) == list:
                k = np.array(k)
            z = np.array([z])
            single_redshift = True
        elif type(k) == float and type(z) != float:
            if type(z) == list:
                z = np.array(z)
            k_val = k
            k = np.array([k_val for _ in range(len(z))])
        elif type(k) != float and type(z) != float:
            if len(k) != len(z):
                raise Exception("Dimensions of k and z do not match:",len(k),len(z))
        return k,z,single_redshift
            
            
        

    def predict(self, k: Union[float,list], z: Union[float,list], f: float = 1.0, vk: float = 0., Gamma: float = 1e-10, p: float = 0.0) -> list:
        
        k,z,single_redshift = self.check_interpret_k_z_input(k,z)
        
        #defaults, do not change
        kmin = 1e-3
        kmax = 5.9
        nsteps_emul = 300
        # k_int = np.logspace(np.log10(kmin),np.log10(kmax),nsteps_emul)
        if max(z)>2.35: raise Exception("Too high redshift encountered (>2.35). Aborting...")
        if vk<0 or vk>5000: raise Exception("Velocity kick outside the training domain [0-5000] km/s. Aborting...")
        if Gamma<0 or Gamma>1/13.5: raise Exception("Decay rate outside the training domain [0-1/13.5] 1/Gyr. Aborting...")
        if f<0.0 or f>1.0: raise Exception("Fraction of 2bDDM outside the training domain [0-1]. Aborting...")

        
        kk = 105
        delta = np.log10(kmax/kmin)/(300-1)
        N = int(1+ np.log10(kk/kmin)/delta)
        k_ext = np.logspace(np.log10(kmin),np.log10(kmin) + (N-1)*delta,N)

        
        PP = []

        par_extend = np.tile(np.array([f,transform_v(vk),transform_g(Gamma)]),(len(z),1))
        params_concat = np.concatenate((par_extend,transform_z(z).reshape(-1,1)),axis = 1)

        
        with torch.no_grad():
            model_out_int_nPca, _ = self.emulator_DDM_2body(torch.tensor(params_concat))
        

        Pks_int = self.t_means + torch.matmul(model_out_int_nPca,self.t_eigenvectors)
        Pks_int = torch.clamp(Pks_int,0.,1.).numpy()

        Pks_ext = np.zeros((len(z),N))
        Pks_ext[:,:nsteps_emul] = Pks_int
        # print('Pk_int.shape,Pk_ext.shape:',Pks_int.shape,Pks_ext.shape)
        """
        p: value
        """
        column = Pks_int[:,-1]+(1-Pks_int[:,-1])*p
        extenstion = np.tile(column.reshape(-1,1),(1,N - nsteps_emul))
        Pks_ext[:,nsteps_emul:] = extenstion

        # for tracking of the extrapolation warning
        warning = False
        for i in range(len(k)):
            if k[i]<= kmin: PP.append(1.0) # for very large scales, boost is always 1.0
            else:
                z_idx = 0 if single_redshift == True else i # only one row in Pks_int and Pks_ext
                if k[i]>kmax: 
                    PP.append(column[z_idx])
                    if not warning: warning = True # extrapolation warning will be thrown before return
                else:
                    idx = bisect.bisect_left(k_ext,k[i])
                    lam = np.log10(k[i]/k_ext[idx-1])/np.log10(k_ext[idx]/k_ext[idx-1])
                    Pk_fill = (1-lam)*Pks_ext[z_idx,idx-1] + lam*Pks_ext[z_idx,idx]
                    PP.append(Pk_fill)
        if warning: print('[emulator.predict] Warning: Extrapolation required! Make sure you know what is hapenning.')
        return np.array(PP)


def transform_v(v: float) -> float:
    vks = [0,200,500,1000,5000]
    tvks = np.log10(np.array(vks)+50)
    tvmin = tvks[0]
    tvks -= tvmin
    tvmax = tvks[-1]
    return (np.log10(v + 50) - tvmin)/tvmax


def transform_g(g: float,a: float =115.396) -> float:
    # return g
    return np.log10(a*g+1)


def transform_z(z: float,a: float =3.97) -> float:
    # return z/zs[0]
    return np.log10(a*z+1)


def check_suppression(line: list) -> list:
    for i in range(len(line)):
        if line[i]>1.00:
            line[i] = 1.00
    return line 
